### PR TITLE
Feature/ir receiver

### PR DESCRIPTION
**Pull request by molitoris**
Hi @h3ll5ur7er 

I implemented the IR receiver found from the [pico examples](https://github.com/raspberrypi/pico-examples/tree/master/pio/ir_nec). It even uses the state machine of the pico :).

I only have a single type of IR receiver (VS1838B) and one IR remote at home. Unfortunately, this combination did not work. I might still have some bugs or the hardware did not work. We can test it on Monday.

I included a blinking LED and serial output to ensure that the device is actually running.

### CODE REVIEW

**Comment by molitoris** [comment]
I finally got the ir receiver to work after soldering the pico with the header pins and using another remote :)
serial output

File: app/src/App.cpp [hunk 1]
```diff
@@ -1,12 +1,60 @@
...
+    gpio_init(LED_PIN);
+    gpio_set_dir(LED_PIN, GPIO_OUT);
+
+    // ir receiver setup
+    PIO pio = pio0;
+    Receiver rec;
+
+    int rx_sm = rec.init(
+        pio,    // programable io
+        15      // gpio input pin
```
Comment (h3ll5ur7er) [inline app/src/App.cpp]: FYI: on the "real" board GPIO15 will be connected to the power control chip, IRRX is connected to GPIO13. We should create a centralized header file with defines to have a single place to store the configuration.

File: app/src/App.cpp [hunk 2]
```diff
@@ -1,12 +1,60 @@
...
-    receiver.SampleDriverFunction();
+
+    const uint8_t LED_PIN = 25;
+
+    // LED setup
+    gpio_init(LED_PIN);
+    gpio_set_dir(LED_PIN, GPIO_OUT);
+
+    // ir receiver setup
+    PIO pio = pio0;
```
Comment (h3ll5ur7er) [inline app/src/App.cpp]: same for pio0/1: we should centeralize the configuration. i think the for now we should pass all IO config via constructor args.

Comment (h3ll5ur7er) [reply app/src/App.cpp]: or as parameter to the init method..

File: app/src/App.cpp [hunk 3]
```diff
@@ -1,12 +1,60 @@
...
+#include "receiver.hpp"
+
 #include "AppHdr.hpp"
 
 
 void ChristmasClock::ChristmasClock::Init() {
-    Receiver receiver;
-    receiver.SampleDriverFunction();
+
+    const uint8_t LED_PIN = 25;
```
Comment (h3ll5ur7er) [inline app/src/App.cpp]: is the LED_PIN used in "production" or is it only for debugging/development purposes? GPIO25 is one of our external GPIOs with clock functionality (ses https://datasheets.raspberrypi.com/rp2040/rp2040-datasheet.pdf page 14)

File: app/src/App.cpp [hunk 4]
```diff
@@ -1,12 +1,60 @@
...
+    );
+    if(rx_sm == -1) {
+        printf("Could not configure PIO\n");
+        return;
+    }
+
+    uint8_t rx_address, rx_data;
+
+    // main loop
+    while(true){
```
Comment (h3ll5ur7er) [inline app/src/App.cpp:17]: we have to coordinate our mainloops, i would suggest to create a update method inside the driver containing one cycle of the loop, which is called from the main mainloop (see [feature/led-driver](https://github.com/h3ll5ur7er/ChristmasClockSDK/compare/feature/led-driver)).

File: app/src/App.cpp (hunk 2 above)
Comment (h3ll5ur7er) [reply app/src/App.cpp]: sorry, i didn't realize this was in App.cpp.. i would suggest to only put the configuration here and move the initialization logic into the driver or at least into a separate method in here to keep the App to the actual business logic

File: drivers/ir/ir.cmake [hunk 5]
```diff
@@ -1,12 +1,18 @@
 set(DRIVER_NAME ir)
 add_library(${DRIVER_NAME} INTERFACE)
 
+# invoke pio_asm to assemble the state machine program
+pico_generate_pio_header(${DRIVER_NAME} ${CMAKE_CURRENT_LIST_DIR}/src/nec_receive.pio)
```
Comment (h3ll5ur7er) [inline drivers/ir/ir.cmake]: put generated code into a "generated" subdirectory to keep people from editing it manually by accident (see [feature/led-driver](https://github.com/h3ll5ur7er/ChristmasClockSDK/compare/feature/led-driver))

**Comment by h3ll5ur7er** [comment]
Very nice work, thank you.. let's discuss my comments later.. i'll be at the office soon..

File: app/src/App.cpp (hunk 3 above)
Comment (molitoris) [reply app/src/App.cpp]: Only for debugging. I removed it.

File: app/src/App.cpp (hunk 4 above)
Comment (molitoris) [reply app/src/App.cpp:17]: I reduce the main loop to only contain the business logic.

File: app/src/App.cpp (hunk 2 above)
Comment (molitoris) [reply app/src/App.cpp]: I kept the config at the top of init. As soon as we have a config file, we can move it there. I reduced the footprint in the app to the minimal business logic.

File: app/src/App.cpp (hunk 1 above)
Comment (molitoris) [reply app/src/App.cpp]: I changed the pins

File: drivers/ir/ir.cmake (hunk 5 above)
Comment (molitoris) [reply drivers/ir/ir.cmake]: Moved it :)

**Comment by molitoris** [comment]
@h3ll5ur7er Thank you for your inputs. I adapted to code accordingly.

**Review by h3ll5ur7er — APPROVED** [review]

**Pull request merged**